In [1]:
#importing necessary libraries

import pandas as pd
import numpy as np
import importlib
import datetime
import functions as fc
import re

importlib.reload(fc)

C:\Users\Sabrina\AppData\Local\Temp\ipykernel_135384\2280367382.py:3: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


<module 'functions' from 'c:\\Users\\Sabrina\\Desktop\\TestDA\\Week_9\\Final_project\\notebooks\\functions.py'>

In [2]:
#importing csv 
video_games_df = pd.read_csv("C:/Users/Sabrina/Desktop/TestDA/Week_9/Final_project/data/raw/Video Games Sales.csv")

#checking df shape
print(f"This is the shape of video_games_df: {video_games_df.shape}")

#visualizing first and last 5 rows
display(video_games_df.head())
display(video_games_df.tail())

This is the shape of video_games_df: (1907, 13)


index  Rank         Game Title Platform    Year     Genre Publisher  \
0      0     1         Wii Sports      Wii  2006.0    Sports  Nintendo   
1      1     2  Super Mario Bros.      NES  1985.0  Platform  Nintendo   
2      2     3     Mario Kart Wii      Wii  2008.0    Racing  Nintendo   
3      3     4  Wii Sports Resort      Wii  2009.0    Sports  Nintendo   
4      4     5             Tetris       GB  1989.0    Puzzle  Nintendo   

   North America  Europe  Japan  Rest of World  Global  Review  
0          40.43   28.39   3.77           8.54   81.12   76.28  
1          29.08    3.58   6.81           0.77   40.24   91.00  
2          14.50   12.22   3.63           3.21   33.55   82.07  
3          14.82   10.51   3.18           3.01   31.52   82.65  
4          23.20    2.26   4.22           0.58   30.26   88.00

index  Rank                        Game Title Platform    Year  \
1902   1902  1903  Lizzie McGuire 2: Lizzie Diaries      GBA  2004.0   
1903   1903  1904              Xenoblade Chronicles      Wii  2010.0   
1904   1904  1905                     SingStar Abba      PS3  2008.0   
1905   1905  1906    FIFA Soccer World Championship      PS2  2000.0   
1906   1906  1907        WWE SmackDown vs. Raw 2011     X360  2010.0   

             Genre                    Publisher  North America  Europe  Japan  \
1902        Action   Disney Interactive Studios           0.60    0.22   0.00   
1903  Role-Playing                     Nintendo           0.39    0.22   0.16   
1904          Misc  Sony Computer Entertainment           0.25    0.44   0.00   
1905        Sports              Electronic Arts           0.27    0.21   0.28   
1906      Fighting                          THQ           0.42    0.32   0.00   

      Rest of World  Global  Review  
1902           0.01    0.83   55.00  
1903           0.07    0.83   91.74  
1904           0.14    0.83   73.00  
1905           0.07    0.83   73.00  
1906           0.09    0.83   82.00

In [3]:
#checking nan values on the df
video_games_df.isna().sum()

index             0
Rank              0
Game Title        0
Platform          0
Year             29
Genre             0
Publisher         2
North America     0
Europe            0
Japan             0
Rest of World     0
Global            0
Review            0
dtype: int64

Luckily, we only have missing values on Year and Publisher, which can be easily fill by isolating the games that correspond to these NaN values and substitude them with the correct year and publisher in order to avoid information lose. 

In [4]:
#checking length is matching the previous info. 
print(len(video_games_df[video_games_df['Year'].isna()]))

#visualizing games with NaN'
nans_year = video_games_df[video_games_df['Year'].isna()]
list_games_nans = nans_year['Game Title'].values.tolist()
print(list_games_nans)

display(video_games_df[video_games_df['Year'].isna()])

29
['Madden NFL 2004', 'RollerCoaster Tycoon', 'FIFA Soccer 2004', 'wwe Smackdown vs. Raw 2006', 'LEGO Batman: The Videogame', 'Rock Band', "Frogger's Adventures: Temple of the Frog", 'LEGO Indiana Jones: The Original Adventures', 'Call of Duty 3', 'Rock Band', 'Gran Turismo Concept 2001 Tokyo', 'Rock Band', 'Triple Play 99', 'Call of Duty: Black Ops', 'NASCAR Thunder 2003', 'Hitman 2: Silent Assassin', 'Dragon Warrior VII', 'LEGO Batman: The Videogame', 'Rock Band', 'Legacy of Kain: Soul Reaver', 'Donkey Kong Land III', 'Suikoden III', 'Wheel of Fortune', 'Teenage Mutant Ninja Turtles', 'Namco Museum', 'LEGO Harry Potter: Years 5-7', 'Blitz: The League', 'Yakuza 4', 'Madden NFL 07']


index  Rank                                   Game Title Platform  Year  \
143     143   144                              Madden NFL 2004      PS2   NaN   
253     253   254                         RollerCoaster Tycoon       PC   NaN   
301     301   302                             FIFA Soccer 2004      PS2   NaN   
373     373   374                   wwe Smackdown vs. Raw 2006      PS2   NaN   
434     434   435                   LEGO Batman: The Videogame      Wii   NaN   
523     523   524                                    Rock Band     X360   NaN   
524     524   525     Frogger's Adventures: Temple of the Frog      GBA   NaN   
573     573   574  LEGO Indiana Jones: The Original Adventures      Wii   NaN   
582     582   583                               Call of Duty 3      Wii   NaN   
646     646   647                                    Rock Band      Wii   NaN   
746     746   747              Gran Turismo Concept 2001 Tokyo      PS2   NaN   
936     936   937                                    Rock Band      PS3   NaN   
1016   1016  1017                               Triple Play 99       PS   NaN   
1030   1030  1031                      Call of Duty: Black Ops       PC   NaN   
1275   1275  1276                          NASCAR Thunder 2003      PS2   NaN   
1318   1318  1319                    Hitman 2: Silent Assassin       XB   NaN   
1326   1326  1327                           Dragon Warrior VII      3DS   NaN   
1386   1386  1387                   LEGO Batman: The Videogame      PSP   NaN   
1424   1424  1425                                    Rock Band      PS2   NaN   
1526   1526  1527                  Legacy of Kain: Soul Reaver       PS   NaN   
1554   1554  1555                         Donkey Kong Land III       GB   NaN   
1631   1631  1632                                 Suikoden III      PS2   NaN   
1676   1676  1677                             Wheel of Fortune      PS2   NaN   
1716   1716  1717                 Teenage Mutant Ninja Turtles      GBA   NaN   
1752   1752  1753                                 Namco Museum       XB   NaN   
1759   1759  1760                 LEGO Harry Potter: Years 5-7      Wii   NaN   
1784   1784  1785                            Blitz: The League      PS2   NaN   
1819   1819  1820                                     Yakuza 4      PS3   NaN   
1899   1899  1900                                Madden NFL 07      PSP   NaN   

             Genre                               Publisher  North America  \
143         Sports                         Electronic Arts           4.26   
253       Strategy                                 Unknown           3.82   
301         Sports                                 Unknown           0.59   
373       Fighting                                     NaN           1.57   
434      Adventure  Warner Bros. Interactive Entertainment           1.59   
523           Misc                         Electronic Arts           1.88   
524      Adventure                                 Unknown           2.15   
573      Adventure                               LucasArts           1.44   
582        Shooter                                 Unknown           1.13   
646           Misc                               MTV Games           1.31   
746         Racing                                 Unknown           0.00   
936           Misc                         Electronic Arts           0.96   
1016        Sports                                     NaN           0.81   
1030       Shooter                              Activision           0.53   
1275        Racing                                 Unknown           0.60   
1318        Action                       Eidos Interactive           0.76   
1326  Role-Playing                             Square Enix           0.00   
1386     Adventure  Warner Bros. Interactive Entertainment           0.55   
1424          Misc                                 Unknown           0.71   
1526        Action                 

In [5]:
#updating year values through function 

fc.update_year(video_games_df, [1676], 1975.0)
fc.update_year(video_games_df, [1554], 1997.0)
fc.update_year(video_games_df, [1016], 1998.0)
fc.update_year(video_games_df, [253, 1526], 1999.0)
fc.update_year(video_games_df, [1326], 2000.0)
fc.update_year(video_games_df, [524], 2001.0)
fc.update_year(video_games_df, [746, 1275, 1318, 1631, 1752], 2002.0)
fc.update_year(video_games_df, [143, 301, 1716], 2003.0)
fc.update_year(video_games_df, [373, 1784], 2005.0)
fc.update_year(video_games_df, [582, 1899], 2006.0)
fc.update_year(video_games_df, [1386, 523, 573, 646, 936, 434, 1424], 2008.0)
fc.update_year(video_games_df, [1030, 1819], 2010.0)
fc.update_year(video_games_df, [1759], 2011.0)

In [6]:
#checking nan values on the df
video_games_df.isna().sum()

index            0
Rank             0
Game Title       0
Platform         0
Year             0
Genre            0
Publisher        2
North America    0
Europe           0
Japan            0
Rest of World    0
Global           0
Review           0
dtype: int64

In [7]:
video_games_df[video_games_df['Publisher'].isna()]

index  Rank                  Game Title Platform    Year     Genre  \
373     373   374  wwe Smackdown vs. Raw 2006      PS2  2005.0  Fighting   
1016   1016  1017              Triple Play 99       PS  1998.0    Sports   

     Publisher  North America  Europe  Japan  Rest of World  Global  Review  
373        NaN           1.57    1.02    0.0           0.41    3.00   91.00  
1016       NaN           0.81    0.55    0.0           0.10    1.46   69.72

In [8]:
video_games_df.iloc[373, video_games_df.columns.get_loc('Publisher')] = 'THQ'
video_games_df.iloc[1016, video_games_df.columns.get_loc('Publisher')] = 'Electronic Arts'

In [9]:
#checking nan values on the df
video_games_df.isna().sum()

index            0
Rank             0
Game Title       0
Platform         0
Year             0
Genre            0
Publisher        0
North America    0
Europe           0
Japan            0
Rest of World    0
Global           0
Review           0
dtype: int64

In [10]:
#cleaning columns names before proceeding
fc.rename_columns(video_games_df)
display(video_games_df)

index  rank                        game_title platform    year  \
0         0     1                        Wii Sports      Wii  2006.0   
1         1     2                 Super Mario Bros.      NES  1985.0   
2         2     3                    Mario Kart Wii      Wii  2008.0   
3         3     4                 Wii Sports Resort      Wii  2009.0   
4         4     5                            Tetris       GB  1989.0   
...     ...   ...                               ...      ...     ...   
1902   1902  1903  Lizzie McGuire 2: Lizzie Diaries      GBA  2004.0   
1903   1903  1904              Xenoblade Chronicles      Wii  2010.0   
1904   1904  1905                     SingStar Abba      PS3  2008.0   
1905   1905  1906    FIFA Soccer World Championship      PS2  2000.0   
1906   1906  1907        WWE SmackDown vs. Raw 2011     X360  2010.0   

             genre                    publisher  north_america  europe  japan  \
0           Sports                     Nintendo          40.43   28.39   3.77   
1         Platform                     Nintendo          29.08    3.58   6.81   
2           Racing                     Nintendo          14.50   12.22   3.63   
3           Sports                     Nintendo          14.82   10.51   3.18   
4           Puzzle                     Nintendo          23.20    2.26   4.22   
...            ...                          ...            ...     ...    ...   
1902        Action   Disney Interactive Studios           0.60    0.22   0.00   
1903  Role-Playing                     Nintendo           0.39    0.22   0.16   
1904          Misc  Sony Computer Entertainment           0.25    0.44   0.00   
1905        Sports              Electronic Arts           0.27    0.21   0.28   
1906      Fighting                          THQ           0.42    0.32   0.00   

      rest_of_world  global  review  
0              8.54   81.12   76.28  
1              0.77   40.24   91.00  
2              3.21   33.55   82.07  
3              3.01   31.52   82.65  
4              0.58   30.26   88.00  
...             ...     ...     ...  
1902           0.01    0.83   55.00  
1903           0.07    0.83   91.74  
1904           0.14    0.83   73.00  
1905           0.07    0.83   73.00  
1906           0.09    0.83   82.00  

[1907 rows x 13 columns]

In [11]:
#checking data type inconsistencies 
video_games_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1907 entries, 0 to 1906
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          1907 non-null   int64  
 1   rank           1907 non-null   int64  
 2   game_title     1907 non-null   object 
 3   platform       1907 non-null   object 
 4   year           1907 non-null   float64
 5   genre          1907 non-null   object 
 6   publisher      1907 non-null   object 
 7   north_america  1907 non-null   float64
 8   europe         1907 non-null   float64
 9   japan          1907 non-null   float64
 10  rest_of_world  1907 non-null   float64
 11  global         1907 non-null   float64
 12  review         1907 non-null   float64
dtypes: float64(7), int64(2), object(4)
memory usage: 193.8+ KB


In [12]:
#casting data type change for year. 

video_games_df['year'] = video_games_df['year'].astype(int)
display(video_games_df)

index  rank                        game_title platform  year  \
0         0     1                        Wii Sports      Wii  2006   
1         1     2                 Super Mario Bros.      NES  1985   
2         2     3                    Mario Kart Wii      Wii  2008   
3         3     4                 Wii Sports Resort      Wii  2009   
4         4     5                            Tetris       GB  1989   
...     ...   ...                               ...      ...   ...   
1902   1902  1903  Lizzie McGuire 2: Lizzie Diaries      GBA  2004   
1903   1903  1904              Xenoblade Chronicles      Wii  2010   
1904   1904  1905                     SingStar Abba      PS3  2008   
1905   1905  1906    FIFA Soccer World Championship      PS2  2000   
1906   1906  1907        WWE SmackDown vs. Raw 2011     X360  2010   

             genre                    publisher  north_america  europe  japan  \
0           Sports                     Nintendo          40.43   28.39   3.77   
1         Platform                     Nintendo          29.08    3.58   6.81   
2           Racing                     Nintendo          14.50   12.22   3.63   
3           Sports                     Nintendo          14.82   10.51   3.18   
4           Puzzle                     Nintendo          23.20    2.26   4.22   
...            ...                          ...            ...     ...    ...   
1902        Action   Disney Interactive Studios           0.60    0.22   0.00   
1903  Role-Playing                     Nintendo           0.39    0.22   0.16   
1904          Misc  Sony Computer Entertainment           0.25    0.44   0.00   
1905        Sports              Electronic Arts           0.27    0.21   0.28   
1906      Fighting                          THQ           0.42    0.32   0.00   

      rest_of_world  global  review  
0              8.54   81.12   76.28  
1              0.77   40.24   91.00  
2              3.21   33.55   82.07  
3              3.01   31.52   82.65  
4              0.58   30.26   88.00  
...             ...     ...     ...  
1902           0.01    0.83   55.00  
1903           0.07    0.83   91.74  
1904           0.14    0.83   73.00  
1905           0.07    0.83   73.00  
1906           0.09    0.83   82.00  

[1907 rows x 13 columns]

In [13]:
#dropping columns index and rank
video_games_df = video_games_df.drop(columns=['index', 'rank'])
display(video_games_df)

game_title platform  year         genre  \
0                           Wii Sports      Wii  2006        Sports   
1                    Super Mario Bros.      NES  1985      Platform   
2                       Mario Kart Wii      Wii  2008        Racing   
3                    Wii Sports Resort      Wii  2009        Sports   
4                               Tetris       GB  1989        Puzzle   
...                                ...      ...   ...           ...   
1902  Lizzie McGuire 2: Lizzie Diaries      GBA  2004        Action   
1903              Xenoblade Chronicles      Wii  2010  Role-Playing   
1904                     SingStar Abba      PS3  2008          Misc   
1905    FIFA Soccer World Championship      PS2  2000        Sports   
1906        WWE SmackDown vs. Raw 2011     X360  2010      Fighting   

                        publisher  north_america  europe  japan  \
0                        Nintendo          40.43   28.39   3.77   
1                        Nintendo          29.08    3.58   6.81   
2                        Nintendo          14.50   12.22   3.63   
3                        Nintendo          14.82   10.51   3.18   
4                        Nintendo          23.20    2.26   4.22   
...                           ...            ...     ...    ...   
1902   Disney Interactive Studios           0.60    0.22   0.00   
1903                     Nintendo           0.39    0.22   0.16   
1904  Sony Computer Entertainment           0.25    0.44   0.00   
1905              Electronic Arts           0.27    0.21   0.28   
1906                          THQ           0.42    0.32   0.00   

      rest_of_world  global  review  
0              8.54   81.12   76.28  
1              0.77   40.24   91.00  
2              3.21   33.55   82.07  
3              3.01   31.52   82.65  
4              0.58   30.26   88.00  
...             ...     ...     ...  
1902           0.01    0.83   55.00  
1903           0.07    0.83   91.74  
1904           0.14    0.83   73.00  
1905           0.07    0.83   73.00  
1906           0.09    0.83   82.00  

[1907 rows x 11 columns]

In [14]:
#there are some Unknown values in the publisher columns, hence I'll replace them with the actual Publisher keeping in mind the platform and year
display(video_games_df[video_games_df['publisher'] == 'Unknown'])


game_title platform  year         genre  \
253                       RollerCoaster Tycoon       PC  1999      Strategy   
286                             WWF SmackDown!       PS  2000      Fighting   
301                           FIFA Soccer 2004      PS2  2003        Sports   
524   Frogger's Adventures: Temple of the Frog      GBA  2001     Adventure   
582                             Call of Duty 3      Wii  2006       Shooter   
746            Gran Turismo Concept 2001 Tokyo      PS2  2002        Racing   
1275                       NASCAR Thunder 2003      PS2  2002        Racing   
1424                                 Rock Band      PS2  2008          Misc   
1631                              Suikoden III      PS2  2002  Role-Playing   
1676                          Wheel of Fortune      PS2  1975          Misc   
1716              Teenage Mutant Ninja Turtles      GBA  2003      Platform   
1752                              Namco Museum       XB  2002          Misc   
1784                         Blitz: The League      PS2  2005        Sports   
1899                             Madden NFL 07      PSP  2006        Sports   

     publisher  north_america  europe  japan  rest_of_world  global  review  
253    Unknown           3.82    0.01   0.00           0.00    3.83   87.00  
286    Unknown           2.01    1.35   0.06           0.16    3.58   81.00  
301    Unknown           0.59    2.36   0.04           0.51    3.49   81.00  
524    Unknown           2.15    0.18   0.00           0.07    2.39   69.00  
582    Unknown           1.13    0.82   0.00           0.23    2.18   69.91  
746    Unknown           0.00    1.10   0.42           0.33    1.84   80.00  
1275   Unknown           0.60    0.46   0.00           0.16    1.22   81.00  
1424   Unknown           0.71    0.30   0.00           0.11    1.11   82.00  
1631   Unknown           0.29    0.23   0.38           0.08    0.98   87.00  
1676   Unknown           0.47    0.36   0.00           0.12    0.95   55.00  
1716   Unknown           0.67    0.25   0.00           0.02    0.93   69.00  
1752   Unknown           0.77    0.11   0.00           0.04    0.91   64.05  
1784   Unknown           0.74    0.03   0.00           0.12    0.89   77.60  
1899   Unknown           0.77    0.03   0.00           0.04    0.83   84.00

In [15]:
video_games_df['publisher'].unique()

array(['Nintendo', 'Take-Two Interactive', 'Microsoft Game Studios',
       'Sony Computer Entertainment', 'Activision', 'Electronic Arts',
       'Ubisoft', 'Red Orb', 'Square', 'Bethesda Softworks', 'Capcom',
       '505 Games', 'GT Interactive', 'Konami Digital Entertainment',
       'Sega', 'Square Enix', 'Virgin Interactive',
       'Universal Interactive', 'Eidos Interactive', 'RedOctane',
       'LucasArts', 'Atari', 'Vivendi Games', 'Enix Corporation',
       'Namco Bandai Games', 'Warner Bros. Interactive Entertainment',
       'Hasbro Interactive', 'Unknown', 'THQ',
       'Disney Interactive Studios', 'Acclaim Entertainment', 'Level 5',
       'MTV Games', 'Codemasters', 'id Software', 'Havas Interactive',
       'Majesco Entertainment', 'Midway Games', 'Hudson Soft', 'JVC',
       'Valve Software', '989 Studios', 'Maxis', 'Square EA',
       'Tecmo Koei', 'Gotham Games', 'Deep Silver', 'Infogrames',
       'Arena Entertainment', 'Mindscape', 'Banpresto', 'NCSoft',
       'C

In [16]:
#searched publisher and make sure change is homogeneous with the existing ones 

video_games_df.iloc[253, video_games_df.columns.get_loc('publisher')] = 'Hasbro Interactive'
video_games_df.iloc[286, video_games_df.columns.get_loc('publisher')] = 'THQ'
video_games_df.iloc[301, video_games_df.columns.get_loc('publisher')] = 'Electronic Arts'
video_games_df.iloc[524, video_games_df.columns.get_loc('publisher')] = 'Konami Digital Entertainment'
video_games_df.iloc[582, video_games_df.columns.get_loc('publisher')] = 'Activision'
video_games_df.iloc[746, video_games_df.columns.get_loc('publisher')] = 'Sony Computer Entertainment'
video_games_df.iloc[1275, video_games_df.columns.get_loc('publisher')] = 'Electronic Arts'
video_games_df.iloc[1424, video_games_df.columns.get_loc('publisher')] = 'Electronic Arts'
video_games_df.iloc[1631, video_games_df.columns.get_loc('publisher')] = 'Konami Digital Entertainment'
video_games_df.iloc[1676, video_games_df.columns.get_loc('publisher')] = 'Atari'
video_games_df.iloc[1716, video_games_df.columns.get_loc('publisher')] = 'Konami Digital Entertainment'
video_games_df.iloc[1752, video_games_df.columns.get_loc('publisher')] = 'Namco Bandai Games'
video_games_df.iloc[1784, video_games_df.columns.get_loc('publisher')] = 'Midway Games'
video_games_df.iloc[1899, video_games_df.columns.get_loc('publisher')] = 'Electronic Arts'

In [20]:
#I wont be doing any changes on these column for now
video_games_df['genre'].unique()
video_games_df['year'].unique()
video_games_df['platform'].unique()

array(['Wii', 'NES', 'GB', 'DS', 'PS2', 'SNES', 'X360', 'GBA', 'PS3',
       'N64', 'PC', 'PS', 'XB', '3DS', 'PSP', 'GC', 'GEN', 'DC', 'SAT',
       'WiiU', 'SCD', 'PSV'], dtype=object)

In [22]:
video_games_df.select_dtypes(include=np.number).describe().round(3)

year  north_america    europe     japan  rest_of_world    global  \
count  1907.000       1907.000  1907.000  1907.000       1907.000  1907.000   
mean   2003.760          1.259     0.707     0.317          0.206     2.489   
std       5.907          1.957     1.149     0.725          0.343     3.563   
min    1975.000          0.000     0.000     0.000          0.000     0.830   
25%    2000.500          0.510     0.230     0.000          0.060     1.110   
50%    2005.000          0.810     0.440     0.020          0.130     1.530   
75%    2008.000          1.375     0.810     0.300          0.220     2.540   
max    2012.000         40.430    28.390     7.200          8.540    81.120   

         review  
count  1907.000  
mean     79.039  
std      10.617  
min      30.500  
25%      74.000  
50%      81.000  
75%      86.230  
max      97.000

In here we can see that the average of the games are from 2003 having an average review of 79%, with higher sells in North America, presenting an amount of 1.26 millions units sold, a 50,58% of the total units sold. 
The standard deviation for millions/units sold is higher in north america, preceeded by europe, japan and rest of the world.
We have a minimum amount of 0 millions/units sold per region, and 0.83 globally. 
While in the maximum amount is set to 81.12 million units globally, having a difference of 98.98% with the minimum global amount of units sold which can lead us to think that games have gain considerably weight throughout the years. 

In [25]:
video_games_df.isna().sum()

game_title       0
platform         0
year             0
genre            0
publisher        0
north_america    0
europe           0
japan            0
rest_of_world    0
global           0
review           0
dtype: int64

In [30]:
video_games_df.drop(columns='game_title').select_dtypes(exclude=np.number).mode()


platform   genre        publisher
0      PS2  Sports  Electronic Arts